core/vcpu.c

```diff
@@ -1956,6 +1956,10 @@ static int vcpu_emulate_insn(struct vcpu_t *vcpu)
     uint64 rip = vcpu->state->_rip;
     uint64 va;
 
+    // Clean up the emulation context of the previous MMIO instruction, so that
+    // even if things go wrong, the behavior will still be predictable.
+    vcpu_init_emulator(vcpu);
```

Comment:
vcpu_init_emulator only needs to be called in EM_ERROR case?

Reply:
If we failed to decode/emulate the previous instruction, we should have terminated the guest already and wouldn't have arrived here. The intention here is to always reset the emulation context (`vcpu->emulate_ctxt`) before we start working on a new instruction, because `emulate_ctxt` is shared among all MMIO instructions executed by this vCPU, and we don't want to accidentally reuse anything from the previous context.

BTW, I have made some changes to this patch, and will upload the new version later on.